src/servers/http_h2o_server.cc

```diff
@@ -0,0 +1,690 @@
+
+  // TODO Add server.crt and key files
+  if (USE_HTTPS && setup_ssl(
+                       "h2o/server.crt", "h2o/server.key",
+                       "DEFAULT:!MD5:!DSS:!DES:!RC4:!RC2:!SEED:!IDEA:!NULL:!"
+                       "ADH:!EXP:!SRP:!PSK") != 0)
+    goto Error;
+
+  accept_ctx.ctx = &ctx;
+  accept_ctx.hosts = config.hosts;
+
+  if (create_listener(port_) != 0) {
+    LOG_VERBOSE(1) << stderr << "failed to listen to 0.0.0.0:" << port_ << ":"
+                   << strerror(errno);
+    goto Error;
+  }
+
+  exit_loop = false;
+  while (!exit_loop) {
+#if H2O_USE_LIBUV
+    uv_run(ctx.loop, UV_RUN_DEFAULT);
+#else
+    h2o_evloop_run(ctx.loop, INT32_MAX);
+#endif
+  }
+
+  return nullptr;
+
+Error:
+  return TRTSERVER_ErrorNew(
+      TRTSERVER_ERROR_INTERNAL, "HTTP h2o server failed to run.");
+}
+
+TRTSERVER_Error*
+HTTPAPIServer::Stop()
+{
+  if (!exit_loop) {
+#if H2O_USE_LIBUV
+    uv_close((uv_handle_t*)&listener, NULL);
+#else
+    h2o_socket_read_stop(listener_socket);
+    h2o_socket_close(listener_socket);
+#endif
+
+    // this will break the event loop
+    exit_loop = true;
+  } else {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_UNAVAILABLE, "HTTP h2o server is not running.");
+  }
+
+  return nullptr;
+}
+
+TRTSERVER_Error*
+HTTPAPIServer::ResponseAlloc(
+    TRTSERVER_ResponseAllocator* allocator, const char* tensor_name,
+    size_t byte_size, TRTSERVER_Memory_Type preferred_memory_type,
+    int64_t preferred_memory_type_id, void* userp, void** buffer,
+    void** buffer_userp, TRTSERVER_Memory_Type* actual_memory_type,
+    int64_t* actual_memory_type_id)
+{
+  return nullptr;  // Success
+}
+
+TRTSERVER_Error*
+HTTPAPIServer::ResponseRelease(
+    TRTSERVER_ResponseAllocator* allocator, void* buffer, void* buffer_userp,
+    size_t byte_size, TRTSERVER_Memory_Type memory_type, int64_t memory_type_id)
+{
+  LOG_VERBOSE(1) << "HTTP h2o release: "
+                 << "size " << byte_size << ", addr " << buffer;
+
+  // Don't do anything when releasing a buffer since ResponseAlloc
+  // wrote directly into the response ebvuffer.
+  return nullptr;  // Success
+}
+
+TRTSERVER_Error*
+HTTPServer::CreateAPIServer(
+    const std::shared_ptr<TRTSERVER_Server>& server,
+    const std::shared_ptr<nvidia::inferenceserver::TraceManager>& trace_manager,
+    const std::shared_ptr<SharedMemoryBlockManager>& smb_manager,
+    const std::map<int32_t, std::vector<std::string>>& port_map, int thread_cnt,
+    std::vector<std::unique_ptr<HTTPServer>>* http_servers)
+{
+  if (port_map.empty()) {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_INVALID_ARG,
+        "HTTP is enabled but none of the service endpoints have a valid "
+        "port assignment");
+  }
+  http_servers->clear();
+  for (auto const& ep_map : port_map) {
+    std::string addr = "0.0.0.0:" + std::to_string(ep_map.first);
+    LOG_INFO << "Starting HTTPService at " << addr;
+    http_servers->emplace_back(new HTTPAPIServer(
```

Comment:
Should reconsider class hierarchy based on what h2o supports. Can we support multiple ports within same object? For handlers can we just register the regexs with h2o and it do all the routing to our health, status, etc. endpoints.

Reply:
We cannot support multiple ports with the same object. Yes we just register the endpoints in h2o and let it do the routing to the different endpoints

Reply:
@deadeyegoodwin unfortunately the does not support regex like input while registering the handler. We will have to use the above approach for now.